### PR TITLE
Add dynamoDbRetry

### PR DESCRIPTION
**Pull request by szabado**

### CODE REVIEW

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.amazonaws.services.dynamodbv2.model.AmazonDynamoDBException
+import java.time.Duration
+import misk.backoff.Backoff
+import misk.backoff.DontRetryException
+import misk.backoff.ExponentialBackoff
+import misk.backoff.retry
+
+object DynamoDbHelpers {
+  val DEFAULT_BACKOFF = ExponentialBackoff(
```
Comment (zhxnlai) [inline misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:11]: Unfortunately this object is [stateful](https://github.com/cashapp/misk/blob/master/misk-core/src/main/kotlin/misk/backoff/ExponentialBackoff.kt#L17). You probably want to instantiate a new one in `dynamoDbRetry`

Comment (szabado) [reply misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:11]: ... I really should have seen that coming. My bad

**Review by mightyguava — COMMENTED** [review]
I'm pretty sure the AWS SDK has built-in retries for DynamoDB.

Also, I actually think that this is something you should define in your application, versus putting in MiSK. How retries are done is an application-level concern. The small amount of code reuse probably doesn't make it worthwhile to stick in the framework.

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt (hunk 1 above)
Comment (zhxnlai) [reply misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt]: nah I think it was just bad API design

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import java.time.Duration
+import misk.backoff.Backoff
+import misk.backoff.DontRetryException
+import misk.backoff.ExponentialBackoff
+import misk.backoff.retry
+
+object DynamoDbHelpers {
+  fun getDefaultBackoff(): ExponentialBackoff {
+    return ExponentialBackoff(
+        baseDelay = Duration.ofMillis(10),
```
Comment (szabado) [inline misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:13]: Do these seem reasonable? They're not based on any hard numbers

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import misk.backoff.Backoff
+import misk.backoff.DontRetryException
+import misk.backoff.ExponentialBackoff
+import misk.backoff.retry
+
+object DynamoDbHelpers {
+  val DEFAULT_BACKOFF = ExponentialBackoff(
+      baseDelay = Duration.ofMillis(10),
+      maxDelay = Duration.ofMillis(500),
+      jitter = Duration.ofMillis(5)
```
Comment (miladghaznavi) [inline misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt]: We may need larger ``jitter`` to decrease the chance of time conflicts, how about a ``jitter`` of 10?

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  )
+
+  /**
+   * Retries any [AmazonDynamoDBException] or [PleaseRetryException]. [DontRetryException] will
+   * not be retried.
+   */
+  fun <T> dynamoDbRetry(
+    retries: Int = 3,
```
Comment (miladghaznavi) [inline misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:24]: why 3?

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    )
+  }
+
+  /**
+   * Retries any [AmazonDynamoDBException] or [PleaseRetryException]. [DontRetryException] will
+   * not be retried.
+   */
+  fun <T> dynamoDbRetry(
+    retries: Int = 3,
+    backoff: Backoff = getDefaultBackoff(),
```
Comment (zhxnlai) [inline misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:25]: I think this is still incorrect: `getDefaultBackoff()` gets evaluated once statically and is shared as a global variable

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import misk.backoff.DontRetryException
+import misk.backoff.ExponentialBackoff
+import misk.backoff.retry
+
+object DynamoDbHelpers {
+  fun getDefaultBackoff(): ExponentialBackoff {
+    return ExponentialBackoff(
+        baseDelay = Duration.ofMillis(10),
+        maxDelay = Duration.ofMillis(500),
+        jitter = Duration.ofMillis(5)
```
Comment (miladghaznavi) [inline misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:15]: We might need larger jitters to reduce the chance of conflicts. Maybe ``10``?

**Comment by szabado** [comment]
@mightyguava the thing that this is designed to help is optimistic locking. The SDK doesn't allow you to retry everything (i.e. fetch the new version and then retry the write).

That said, I totally understand on the code reuse front. I didn't realize `retry` existed until I wrote this, that covers the hard part of this and the rest isn't the most reusable block of code

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt (hunk 2 above)
Comment (zhxnlai) [reply misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:13]: I would put `100ms`

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt (hunk 4 above)
Comment (szabado) [reply misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:24]: That's the default max in the hibernate transacter

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt (hunk 2 above)
Comment (miladghaznavi) [reply misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:13]: Not sure how trivial is coming up with a good number.

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt (hunk 6 above)
Comment (zhxnlai) [inline misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:15]: I would put a larger value here like 200ms. See https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt (hunk 5 above)
Comment (szabado) [reply misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:25]: I don't follow. Isn't `getDefaultBackoff()` called every time `dynamoDbRetry` is called? Are the default value of parameters static?

**Comment by szabado** [comment]
If there's interest in merging this I'll write tests, if the sentiment from the other maintainers is aligned with Yunchi's then I'll close this

**Comment by mightyguava** [comment]
> That said, I totally understand on the code reuse front. I didn't realize `retry` existed until I wrote this, that covers the hard part of this and the rest isn't the most reusable block of code

Ack. Please also use https://github.com/michaelbull/kotlin-retry instead of misk retry

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import misk.backoff.DontRetryException
+import misk.backoff.ExponentialBackoff
+import misk.backoff.retry
+
+object DynamoDbHelpers {
+  fun getDefaultBackoff(): ExponentialBackoff {
+    return ExponentialBackoff(
+        baseDelay = Duration.ofMillis(100),
+        maxDelay = Duration.ofSeconds(10),
+        jitter = Duration.ofMillis(200)
```
Comment (miladghaznavi) [inline misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:15]: These numbers might lead to connection timeout. All said, I am not sure how can we judge good numbers here. Maybe using default values from other places might make sense

File: misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt (hunk 4 above)
Comment (miladghaznavi) [reply misk-aws/src/main/kotlin/misk/dynamodb/DynamoDbHelpers.kt:24]: sounds reasonable

**Comment by szabado** [comment]
I'm going to close this based on the feedback.

**Comment by lukeSquared** [comment]
Just to summarize, I think the takeaway here is to use [kotlin-retry](https://github.com/michaelbull/kotlin-retry) to wrap the `save` call? @mightyguava do you have any recommendations for the values to use for the retry count, delay, etc.? For context, this PR was prompted by us failing to retry save operations that fail due to optimistic locking exceptions.

**Pull request closed without merging**